### PR TITLE
feat: form state

### PR DESCRIPTION
**Pull request by Robbert**

### CODE REVIEW

File: src/form/state/FormState.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { FormState } from './FormState.model';
+
+export const createFormState = ({ dirty, invalid }: Partial<FormState>) => ({
```
Comment (Yolijn) [inline src/form/state/FormState.ts:3]: What do you think about using default values in the destructuring instead? 

```js
export const createFormState = ({ dirty = false, invalid = false }: Partial<FormState>) => ({
  dirty,
  invalid
})
```

File: src/form/state/FormFieldDeclaration.model.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export interface FormFieldDeclaration {
+  fieldType?: 'checkbox' | 'checkboxgroup' | 'date' | 'datetime' | 'input' | 'radiogroup' | 'select' | 'time';
```
Comment (Yolijn) [inline src/form/state/FormFieldDeclaration.model.ts:2]: Did you not include a single `radiobutton` by design, as radio options should only be used if there's 2 or more?

File: src/form/state/FormFieldDeclaration.test.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+      expect(field.fieldType).toBe('checkboxgroup');
+    });
+
+    it('can be a listbox', () => {
+      const field = createDeclaration({
+        fieldType: 'select',
+      });
+
+      // TODO: "listbox" or "select"?
```
Comment (Yolijn) [inline src/form/state/FormFieldDeclaration.test.ts:74]: Should we determine this before merging?

File: src/form/state/FormFieldDeclaration.model.ts (hunk 2 above)
Comment (Robbert) [reply src/form/state/FormFieldDeclaration.model.ts:2]: Indeed a single checkbox is allowed, whereas a single radio button would not be useful.

Comment (Robbert) [reply src/form/state/FormFieldDeclaration.model.ts:2]: I do think `textarea` is probably missing here.

File: src/form/state/FormFieldDefinition.test.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import { FormFieldDefinition } from './FormFieldDefinition.model';
+
+describe('form field description', () => {
+  describe('display and interaction with value', () => {
+    it.todo('can define the value may be treated as case insensitive');
+
+    it.todo('can define the value must be treated as case sensitive');
+
+    it('can define the value can contain multiple lines of text', () => {
```
Comment (Yolijn) [inline src/form/state/FormFieldDefinition.test.ts:9]: `can define the value can contain` => `can define the value may contain` might read easier without losing it's meaning?

File: src/form/state/FormFieldDefinition.test.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    it('can define the value should not receive spell checking suggestions', () => {
+      const state: FormFieldDefinition = {
+        spellCheck: false,
+      };
+
+      expect(state.spellCheck).toBe(false);
+    });
+
+    it('can define the can contain significant whitespace that must be preserved', () => {
```
Comment (Yolijn) [inline src/form/state/FormFieldDefinition.test.ts:25]: `it('can define the can contain significant whitespace that must be preserved'` => `it('can define the value may contain significant whitespace that must be preserved'`

File: src/form/state/FormFieldInputState.test.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
+import { createDefaultState } from './FormFieldDefaultState';
+import { createInputState } from './FormFieldInputState';
+import { FormFieldInputState } from './FormFieldInputState.model';
+
+describe('Form field input state', () => {
+  describe('value', () => {
+    it('is initially empty', () => {
```
Comment (Yolijn) [inline src/form/state/FormFieldInputState.test.ts:7]: Can we make this `is empty by default` instead as you can actually predefine a value while initiating?

File: src/form/state/FormFieldInputState.test.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+          value,
+        }),
+      );
+
+      expect(inputState.value).toBe(value);
+    });
+  });
+
+  describe('validation state', () => {
+    it('is initially valid', () => {
```
Comment (Yolijn) [inline src/form/state/FormFieldInputState.test.ts:26]: Same as before, can we use `is valid by default` instead? 🙂

File: src/data/AdresRegelData.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  pattern: teletex,
+};
+
+/**
+ * Regel 3 adres buitenland
+ *
+ * Logisch Ontwerp GBA 3.14:
+ *
+ * - elementnummer 13.30
+ * - lengte 1-35
```
Comment (Yolijn) [inline src/data/AdresRegelData.ts:40]: Hmm vooral voor mijn begrip: betekend lengte 1-35 ook gelijk dat hij verplicht is? Of dat als je hem invult de lengte tenminste 1 moet zijn?

File: src/data/GeboorteplaatsData.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { FormFieldDefinition } from '../form/state/FormFieldDefinition.model';
+import { teletex } from './TeletexData';
+
+export const placeOfBirthValidation: FormFieldDefinition = {
+  maxLength: 40,
```
Comment (Yolijn) [inline src/data/GeboorteplaatsData.ts:5]: Hmmm, this is probably set by some registry? People born in the town with the longest name will now not be able to fill in our form 😅 https://en.wikipedia.org/wiki/Taumatawhakatangi%C2%ADhangakoauauotamatea%C2%ADturipukakapikimaunga%C2%ADhoronukupokaiwhen%C2%ADuakitanatahu

File: src/form/dutch-form-fields.test.ts [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+          id,
+          label: 'KVK-nummer',
+          fieldType: 'input',
+          inputSubtype: 'text',
+        },
+        definition: {
+          ...kvkValidation,
+        },
+      });
+      expect(def).not.toBeUndefined();
```
Comment (Yolijn) [inline src/form/dutch-form-fields.test.ts:340]: Hmmm should this maybe be `const {definition} = createFormField({...}); expect(definition).not.toBeUndefined()`?

File: src/form/state/FormFieldDeclaration.test.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+describe('Form control of a form field', () => {
+  describe('form control types', () => {
+    it('can be a textbox', () => {
+      const field = createDeclaration({
+        fieldType: 'input',
+      });
+
+      // TODO: We probably need a better name that `input`
+      expect(field.fieldType).toBe('input');
```
Comment (Yolijn) [inline src/form/state/FormFieldDeclaration.test.ts:18]: Hmmm maybe `textinput` and `textarea`? Should we also have input types for `email` etc.?

Comment (Yolijn) [reply src/form/state/FormFieldDeclaration.test.ts:18]: Or should the `input` be for text, numeric and email inputs together?

File: src/form/state/FormFieldDefaultState.test.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import { createDefaultState } from './FormFieldDefaultState';
+import { FormFieldDefaultState } from './FormFieldDefaultState.model';
+
+describe('Form field default state', () => {
+  describe('value', () => {
+    it('is initially empty', () => {
```
Comment (Yolijn) [inline src/form/state/FormFieldDefaultState.test.ts:6]: `it('is empty by default')`?

File: src/form/state/FormFieldOption.model.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Option, for example in radio group, checkbox group, listbox or
+ * even for autocomplete options in <datalist>
+ */
+export interface FormFieldOption {
+  id: string;
+  label?: string;
+  labelKey?: string;
```
Comment (Yolijn) [inline src/form/state/FormFieldOption.model.ts:8]: 🤔 what is a labelKey?

File: src/form/state/FormSubmitState.test.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
+import { createSubmitState } from './FormSubmitState';
+
+describe('Form submit action state', () => {
+  describe('busy', () => {
+    it('is initially not busy', () => {
```
Comment (Yolijn) [inline src/form/state/FormSubmitState.test.ts:5]: `is not busy by default`?

File: src/data/AdresRegelData.ts (hunk 8 above)
Comment (Robbert) [reply src/data/AdresRegelData.ts:40]: I see your point, but I wouldn't think so. The GBA doc doesn't seem to specify optional fields, even when logically they should be (huisletter, for example). So I would interpret this as: they should either be optional or adhere to this minimum length.

File: src/form/dutch-form-fields.test.ts (hunk 10 above)
Comment (Robbert) [reply src/form/dutch-form-fields.test.ts:340]: These tests are not really worked out to test anything specific, they mostly test they don't throw an error.

File: src/data/GeboorteplaatsData.ts (hunk 9 above)
Comment (Robbert) [reply src/data/GeboorteplaatsData.ts]: I moved everything into a `nl-NL/` directory to make clear these are not intended as international defaults.

**Pull request merged**